src/context.rs

```diff
@@ -88,6 +92,137 @@ impl VerifierContext<circuit::v1_2::CircuitImpl, circuit::v1_2::recursive::Circu
     }
 }
 
+#[derive(Clone, PartialEq, Debug)]
+/// The segment info.
+pub struct SegmentInfo {
+    /// Hash function name
+    pub hashfn: String,
+    /// The power of 2 proof size
+    pub po2: u32,
+}
+
+impl SegmentInfo {
+    /// Create a new segment info
+    pub fn new(hashfn: String, po2: u32) -> Self {
+        Self { hashfn, po2 }
+    }
+}
+
+/// Dynamic verifier trait. It's implemented by all verifier context and can be
+/// used with dynamic dispatching. Expose just the functionalities that can be
+/// dispatched dynamically.
+///
+pub trait Verifier {
+    /// Verify the proof against this verifier context, the given `image_id` and journal.
+    fn verify(
+        &self,
+        image_id: Digest,
+        proof: Proof,
+        pubs: Journal,
+    ) -> Result<(), VerificationError>;
+    fn extract_composite_segments_info(
+        &self,
+        composite: &CompositeReceipt,
+    ) -> Result<alloc::vec::Vec<SegmentInfo>, VerificationError> {
+        composite
+            .segments
+            .iter()
+            .map(|s| {
+                self.extract_segment_info(s.seal.as_slice(), s.hashfn.as_str())
+                    .map(|po2| SegmentInfo {
+                        hashfn: s.hashfn.clone(),
+                        po2,
+                    })
+            })
+            .collect()
+    }
+
+    fn suites(&self) -> &BTreeMap<String, HashSuite<BabyBear>>;
+
+    fn set_suites(&mut self, suites: BTreeMap<String, HashSuite<BabyBear>>);
+
+    fn set_poseidon2_mix_impl(
+        &mut self,
+        poseidon2: alloc::boxed::Box<dyn Poseidon2Mix + Send + Sync + 'static>,
+    );
+
+    fn extract_segment_info(&self, seal: &[u32], hashfn: &str) -> Result<u32, VerificationError>;
+}
+
+impl<SC: CircuitCoreDef, RC: CircuitCoreDef> Verifier for VerifierContext<SC, RC> {
+    fn verify(
+        &self,
+        image_id: Digest,
+        proof: Proof,
+        journal: Journal,
+    ) -> Result<(), VerificationError> {
+        proof.verify(self, image_id, journal.digest())
+    }
+
+    fn suites(&self) -> &BTreeMap<String, HashSuite<BabyBear>> {
+        &self.suites
+    }
+
+    fn set_suites(&mut self, suites: BTreeMap<String, HashSuite<BabyBear>>) {
+        self.suites = suites;
+    }
+    fn set_poseidon2_mix_impl(
+        &mut self,
+        poseidon2: alloc::boxed::Box<dyn Poseidon2Mix + Send + Sync + 'static>,
+    ) {
+        Self::set_poseidon2_mix_impl(self, poseidon2)
+    }
+
+    fn extract_segment_info(&self, seal: &[u32], hashfn: &str) -> Result<u32, VerificationError> {
+        let suite = self
+            .suites
+            .get(hashfn)
+            .ok_or(VerificationError::InvalidHashSuite)?;
+        // Make IOP
+        let mut iop = ReadIOP::<BabyBear>::new(seal, suite.rng.as_ref());
+        let slice: &[BabyBearElem] = iop.read_field_elem_slice(SC::OUTPUT_SIZE + 1);
```

Comment:
Since here `ReadIOP` instance is only used for `read_field_elem_slice`, and not for generating challenges, the result is actually independent of the particular `hashfn`. So `hashfn` could be removed from the interface, and `iop` instantiated with an arbitrary rng. We could implement the `RngFactory` and `Rng` traits for a `DummyRng` struct to make this clear, but I don't think this is really worth it.
More important, I think `extract_segment_po2` could be a better name for the method

Reply:
I've just renamed. You're right that I can just use a dummy rng here, but It will make the code a little bit more confusing.